### PR TITLE
feature/houston-scraper

### PR DESCRIPTION
**Pull request by evamaxfield**
### Description of Changes

_Include a description of the proposed changes._

This is the manual start to the Houston scraper from Joy.

* Manual mapping of Houston data -> CDP Model
* Next up is programmatic scraper

### CODE REVIEW

File: manual.py [hunk 1]
```diff
@@ -0,0 +1,568 @@
...
+        ),
+        ingestion_models.EventMinutesItem(
+            minutes_item = ingestion_models.MinutesItem("ROLL CALL AND ADOPT THE MINUTES OF THE PREVIOUS MEETING")
+        ),
+        ingestion_models.EventMinutesItem(
+            minutes_item = ingestion_models.MinutesItem("MAYOR'S REPORT")
+        ),
+        # where should i put ACCEPT WORK???? following two belongs to accept work but not matters
+        ingestion_models.EventMinutesItem(
+            minutes_item = ingestion_models.MinutesItem("RECOMMENDATION from Director Houston Public Works for approval of final contract amount of $2,949,927.38 and acceptance of work on contract with REYTEC CONSTRUCTION RESOURCES, INC for FY2018 Drainage Rehab Work Orders #4 - 1.67% under the original Contract Amount - DISTRICTS A - PECK; B - JACKSON; C - KAMIN; D - EVANS-SHABAZZ; E - MARTIN; G - HUFFMAN; H - CISNEROS; I - GALLEGOS and K - CASTEX-TATUM")
```
Comment (evamaxfield) [inline manual.py:26]: I figured out a method to get the name and the title of matters.

The name should be: "HPW 20CW10123 / Disclaimer of Harrisburg Alley/SY22-092" its the first text after the item creation date after clicking the link. These are usually pretty short. So I would use that as the name

The title is then the text you have here.

File: manual.py [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+            minutes_item = ingestion_models.MinutesItem("RECOMMENDATION from Director Houston Public Works for approval of final contract amount of $2,979,476.73 and acceptance of work on contract with GRAVA LLC for FY2019 Open Drainage System Maintenance Work Orders #1 - 0.68% under the original Contract Amount - DISTRICTS A - PECK; B - JACKSON; C - KAMIN; D - EVANS-SHABAZZ; E - MARTIN; G - HUFFMAN; H - CISNEROS; I - GALLEGOS; J - POLLARD and K - CASTEX-TATUM")
+        ),
+        ingestion_models.EventMinutesItem(
+            minutes_item = ingestion_models.MinutesItem("PROPERTY - NUMBER 3")
+        ),
+        ingestion_models.EventMinutesItem(
+            minutes_item = ingestion_models.MinutesItem("PURCHASING and TABULATION OF BIDS - NUMBERS 4 through 9")
+        ),
+        ingestion_models.EventMinutesItem(
+            minutes_item = ingestion_models.MinutesItem("ORDINANCES - NUMBERS 10 through 26"),
```
Comment (evamaxfield) [inline manual.py:38]: As mentioned yesterday, these are simply just headings not minutes items or matters. Each of the sub items are the minutes items and matters discussed.

File: manual.py [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+            matter = #How to add more matters
+                ingestion_models.Matter(
+                "ORDINANCE approving and authorizing third amendment to contract between City of Houston and BRENTWOOD ECONOMIC COMMUNITY DEVELOPMENT CORPORATION, d/b/a BRENTWOOD COMMUNITY FOUNDATION, to extend term of contract and provide additional Housing Opportunities for Persons With AIDS Funds for the continuing administration and operation of a Community Residence and an Emergency Rental Assistance Program with Supportive Services - $892,634.00 - Grant Fund - DISTRICT K - CASTEX-TATUM",
```
Comment (evamaxfield) [inline manual.py:41]: As mentioned above, the name is the short text from the matter page near the top

File: manual.py [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+            ingestion_models.Vote(
+                person = ingestion_models.Person(
+                    "Amy Peck",
+                    is_active = "true",
+                    # only has seat/role, what to do when only has role??
+                    seat = ingestion_models.Seat(
+                        "District A"
+                    )
+                ),
+                decision = "AYES"
```
Comment (evamaxfield) [inline manual.py:57]: Where did you find these? I can't find them.

File: manual.py [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+                # title??
+                title = "ORDINANCE approving and authorizing Subrecipient Agreement between City of Houston and CATHOLIC CHARITIES OF THE ARCHDIOCESE OF GALVESTON-HOUSTON to provide Emergency Solutions Grant (ESG), and Emergency Solutions Grant Coronavirus (ESG-CV) Funds for administration and related services in connection with the City’s Emergency Rental Assistance Program, for households that have been affected by COVID-19 - $3,058,423.79 - Grant Fund"
+                ),
+            decision = "ADOPTED",
+            votes = [
+            ingestion_models.Vote(
+                person = ingestion_models.Person(
+                    "Amy Peck",
+                    is_active = "true",
+                    # only has seat/role, what to do when only has role??
```
Comment (evamaxfield) [inline manual.py:52]: Seat information is good. I found their role info on a different page but we may need to manually add that in by basically having a static JSON file to store all that info and the scraper looks up that info by the persons name for example. The page I found their roles is here: http://www.houstontx.gov/council/a/index.html

Or here: http://www.houstontx.gov/council/committees/

The second one looks decently easy to scrape.

But now I am a bit confused. Do they only record full council meetings?

**Review by evamaxfield — COMMENTED** [review]
Going to stop there for now. Good job! I left a bunch of comments to look through and would love if you would respond if you have any questions or comments.

I think next step for you would be to start a scraper.

For houston, I believe you can get away with normal `requests` and `beautifulsoup4` scraping.

Here is a good tutorial: https://www.geeksforgeeks.org/implementing-web-scraping-python-beautiful-soup/

And before you go too deep into scraping there are two things that I think would be good.

First: make a plan, written in comments of what the scraper should do when all complete. Remember the end goal is to get to a single function:

```python
def get_events(start_dt: datetime, end_dt: datetime) -> List[EventIngestionModel]:
```

So under that system, the finished scraper will do:
1. find meetings that match that datetime range
2. for each meeting, open up the marked agenda
   a. find the basic body + session info
   b. parse the minutes items and matters
   ...

That is really rough, you should fill it out with things like "for each matter open up a new page" or whatever you think you will need to do. Just write it out in plain text.

The second thing you should do is start writing the scraper itself. but start small. Maybe write a simple function that when given a url to any of the meetings, returns back the ingestion models for Session and Body. We want to work iteratively so it's easier to comment and easier to revert back in the commits if something goes wrong.

Hope this makes sense. If not, happy to chat and explain more.

**Comment by evamaxfield** [comment]
cc @Joy713

File: manual.py (hunk 4 above)
Comment (Joy713) [reply manual.py:57]: I just realized I forgot to leave you comments about checking where to find the votes info. In terms of the votes info I included in the manual.py you asked about in the comment, I found it in the legal minutes on this website https://houston.novusagenda.com/agendapublic/, which states that all voting aye, so I just put down all the members listed.

File: example-cdp-event.json [hunk 6]
```diff
@@ -0,0 +1,1115 @@
+{
```
Comment (evamaxfield) [inline example-cdp-event.json:1]: Delete this file!

File: manual.py [hunk 7]
```diff
@@ -0,0 +1,569 @@
+from cdp_backend.pipeline import ingestion_models
```
Comment (evamaxfield) [inline manual.py:1]: Delete this file!

File: cdp_scrapers/instances/houston.py [hunk 8]
```diff
@@ -0,0 +1,531 @@
...
+import requests
+from cdp_backend.pipeline import ingestion_models
+from datetime import datetime, timedelta
+from typing import List, Union
+import bs4
+
+# from typing import Tuple, Optional
+
+
+# def get_role(name: str) -> Tuple[List[str], bool]:
```
Comment (evamaxfield) [inline cdp_scrapers/instances/houston.py:11]: leave a comment that says: "Houston has no person info currently but these functions are ready whenever we do have person data"

File: cdp_scrapers/instances/houston.py [hunk 9]
```diff
@@ -0,0 +1,531 @@
...
+    ----------------
+    event: Union[Tag, NavigableString, None]
+        All elements in the page that we want to scrape
+
+    Returns:
+    ----------------
+    str
+        The body name
+    """
+    print("start get body name")
```
Comment (evamaxfield) [inline cdp_scrapers/instances/houston.py:154]: change this to a `log.info("...")` statement!

File: cdp_scrapers/instances/houston.py [hunk 10]
```diff
@@ -0,0 +1,531 @@
...
+    event = remove_extra_type(event)
+    bodyTable = event.find_all("table")[1].find("table")
+    if "CITY COUNCIL" in bodyTable.text:
+        return "City Council"
+    else:
+        return bodyTable.find_all("span")[3].text
```
Comment (evamaxfield) [inline cdp_scrapers/instances/houston.py:160]: You dont need this if else you just need to call

`bodyTable.text.title()` to make it "title case"

File: cdp_scrapers/instances/houston.py [hunk 11]
```diff
@@ -0,0 +1,531 @@
...
+    matter_title = (
+        remove_extra_type(matter.find("table"))
+        .find_all("table")[2]
+        .text.replace("Summary:", "")
+        .strip()
+    )
+    return matter_title
+
+
+def get_eventMinutesItem(
```
Comment (evamaxfield) [inline cdp_scrapers/instances/houston.py:219]: Make sure all functions are in snake case

`blah_balj_blah`

File: cdp_scrapers/instances/houston.py [hunk 12]
```diff
@@ -0,0 +1,531 @@
...
+                                break
+                        if "MATTERS HELD" not in one_minute_type:
+                            event_minutes_items.append(
+                                ingestion_models.EventMinutesItem(
+                                    minutes_item=ingestion_models.MinutesItem(
+                                        get_matter_name(minutes_link)
+                                    )
+                                )
+                            )
+    print("end get items")
```
Comment (evamaxfield) [inline cdp_scrapers/instances/houston.py:359]: Reminder change to `log.info`

File: cdp_scrapers/instances/houston.py [hunk 13]
```diff
@@ -0,0 +1,531 @@
...
+    return event_minutes_items
+
+
+# Big Functions
+main_URL = "https://houstontx.new.swagit.com/views/408"
+main_page = requests.get(main_URL)
+main = BeautifulSoup(main_page.content, "html.parser")
```
Comment (evamaxfield) [inline cdp_scrapers/instances/houston.py:366]: Place these function calls inside the `get_events` function. Don't call / process anything outside of functions!

File: cdp_scrapers/instances/houston.py [hunk 14]
```diff
@@ -0,0 +1,531 @@
...
+    main_tbody = remove_extra_type(main_table.find("tbody"))
+    main_year = main_tbody.find_all("tr")
+    link = ""
+    for year in main_year:
+        cells = year.find_all("td")
+        date = cells[1].text.replace(",", "").strip()
+        date = datetime.strptime(date, "%b %d %Y").date()
+        if date == time:
+            link_post = cells[3].find("a")["href"]
+            link = "https://houstontx.new.swagit.com/" + link_post
```
Comment (evamaxfield) [inline cdp_scrapers/instances/houston.py:420]: Use formatted strings:

`f"https://houstontx.new.swagit.com/{link_post}"`

**Review by evamaxfield — COMMENTED** [review]
Looks really good! A few things to change but everything else looks great.

**Comment by evamaxfield** [comment]
/test-scraper get_houston_events --from_dt=2021-12-01T00:00:00 --to_dt=2021-12-14T00:00:00

**Comment by evamaxfield** [comment]
Also, can you add a snippet in the `__init__.py` file of the `instances` module that first imports your `get_events` function

like so:
`from cdp_scrapers.instances.houston import get_events as get_houston_events`

second, at the bottom of that same file, add `get_houston_events` to the `SCRAPER_FUNCTIONS` dictionary

`SCRAPER_FUNCTIONS["houston"] = get_houston_events`

File: cdp_scrapers/instances/houston.py [hunk 15]
```diff
@@ -0,0 +1,531 @@
+from bs4 import BeautifulSoup, Tag, NavigableString
+import requests
+from cdp_backend.pipeline import ingestion_models
+from datetime import datetime, timedelta
+from typing import List, Union
+import bs4
+
```
Comment (evamaxfield) [inline cdp_scrapers/instances/houston.py:7]: Add a logger so you can call `log.info`

```
import logging

log = logging.getLogger(__name__)
```

**Comment by Joy713** [comment]
/test-scraper get_houston_events --from_dt=2021-12-01T00:00:00 --to_dt=2021-12-14T00:00:00

**Comment by evamaxfield** [comment]
/test-scraper get_houston_events --from_dt=2022-02-01T00:00:00 --to_dt=2022-02-16T00:00:00

**Comment by evamaxfield** [comment]
/test-scraper get_houston_events --from_dt=2022-02-01T00:00:00 --to_dt=2022-02-16T00:00:00

**Pull request closed without merging**